### PR TITLE
sclang: protect keysValuesArrayDo primitive

### LINKED ISSUE #852: sclang crashes on evaluation of private method (due to bug in keysValuesArrayDo)
tested on OS X 10.6.8 & SC v3.4.4 - v3.6.5:

evaluating
o = [Event.prNew]
kills sclang.
Same applies to some other objects, some like Array apparently don't crash.

see also:
http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/sclang-perishes-tp7598889p7598908.html

### PR DESCRIPTION
**Pull request by telephon**
This can happen in corner cases. Fixes #852.

### CODE REVIEW

**Comment by telephon** [comment]
It fails with a double error message:

```
ERROR: primitive failed: keysValuesArrayDo first argument should be an array.
ERROR: Primitive '_ObjectString' failed.
Failed.
RECEIVER:
Instance of Event {    (0x11a59da68, gc=74, fmt=00, flg=00, set=03)
```

But maybe someone can give me hints of how to improve it!

File: lang/LangSource/PyrInterpreter3.cpp [hunk 1]
```diff
@@ -1758,6 +1758,18 @@ HOT void Interpret(VMGlobals *g)
 					PyrSlot * vars = g->frame->vars;
 					int m = slotRawInt(&vars[3]);
 					PyrObject * obj = slotRawObject(&vars[1]);
+					
+					if (IsNil(&vars[1])) {
+						error("primitive failed: keysValuesArrayDo first argument should be an array.\n");
```
Comment (brianlheim) [inline lang/LangSource/PyrInterpreter3.cpp]: Can this error message please match the others in this file? See above line 1666:
```
error("Integer-forBy : endval or stepval not an Integer.\n");
```
and below line 1967:
```
error("Number-forSeries : first, second or last not an Integer or Float.\n");
```

File: lang/LangSource/PyrInterpreter3.cpp [hunk 2]
```diff
@@ -1758,6 +1758,18 @@ HOT void Interpret(VMGlobals *g)
 					PyrSlot * vars = g->frame->vars;
 					int m = slotRawInt(&vars[3]);
 					PyrObject * obj = slotRawObject(&vars[1]);
+					
+					if (IsNil(&vars[1])) {
```
Comment (brianlheim) [inline lang/LangSource/PyrInterpreter3.cpp]: Shouldn't this check happen before lines 1759-1760?

**Comment by brianlheim** [comment]
That error message matches what happens when hitting similar cases in `PyrInterpreter3`:

```supercollider
3.forBy("string"); 
// ERROR: Integer-forBy : endval or stepval not an Integer.
// ERROR: Primitive '_ObjectCompileString' failed.

3.forSeries("garbage");
// ERROR: Number-forSeries : first, second or last not an Integer or Float.
// ERROR: Primitive '_ObjectCompileString' failed.
```

I think it's fine for the purposes of this PR. Thanks for this!

**Comment by brianlheim** [comment]
There's an inconsistency between the check (`IsNil`) and the error message (`should be an array`). One of those needs to be modified—I'd go for more robust type-checking at this point, but since the crash only happens when the input is nil, amending the error message wouldn't be bad either.

**Comment by telephon** [comment]
the check could be more elaborate, but given that this is an internal method, checking against nil is ok, I think. Alternatively, one can check how it's done in the array primitives - it is a little more complicated there.

**Comment by telephon** [comment]
I still find it weird that the second error message is `ERROR: Primitive '_ObjectString' failed.`. But, well.

**Comment by snappizz** [comment]
Can we get a unit test please?

**Comment by brianlheim** [comment]
@telephon Can you give the code you used to produce that error message? I find it interesting that mine refers to the primitive `_ObjectCompileString`.

The code looks pretty good; I am OK to merge once there's a unit test. :)

**Comment by telephon** [comment]
`().keysValuesArrayDo()`

I'll do a test next week.

**Comment by telephon** [comment]
this code: `e = (); e.array = nil; e.keysValuesDo({});` causes the interpreter go in an infinite print loop, repeating:

```
Instance of PrimitiveFailedError {    (0x11a441f88, gc=88, fmt=00, flg=00, set=03)
  instance variables [5]
    what : "Failed."
    protectedBacktrace : nil
    path : nil
    receiver : instance of Event (0x16bbc6fa8, size=5, set=3)
    failedPrimitiveName : Symbol '_ObjectString'
}
ERROR: Dictionary-keysValuesArrayDo: first argument should not be nil.
ERROR: throw during error handling!

```

**Comment by telephon** [comment]
I'm not sure if I'm doing the right thing to the stack here:

```
if (IsNil(&vars[1])) {
						error("Dictionary-keysValuesArrayDo: first argument should not be nil.\n");
						
						slotCopy(++sp, &g->receiver);
						numArgsPushed = 1;
						selector = gSpecialSelectors[opmPrimitiveFailed];
						slot = sp;
						
						goto class_lookup;
					}
```

**Comment by telephon** [comment]
Ok, setting `numArgsPushed = 0` fixes it. Please have a look if that is correct.

**Comment by telephon** [comment]
(can be merged IMHO)

**Comment by brianlheim** [comment]
I was wary of this one mainly because of the uncertain copy-pasting into the main interpreter loop, but I guess since the others seem to be similarly weird with posting `_ObjectCompileString` / `_ObjectString` it's probably OK. One thing I noticed just now is that the unit tests test an Event object and not a Dictionary. I'm going to change that and then merge. Thanks!

**Comment by brianlheim** [comment]
Thanks! Eventually we should get to the bottom of why `_ObjectString` is the named primitive in the error message/call stack.

**Pull request merged**